thriftbp/ttl_client.go

```diff
@@ -19,26 +25,50 @@ var _ Client = (*ttlClient)(nil)
 
 // ttlClient is a Client implementation wrapping thrift's TTransport with a TTL.
 type ttlClient struct {
-	thrift.TClient
+	// configs needed for refresh to work
+	generator ttlClientGenerator
+	ttl       time.Duration
+	jitter    float64
 
-	transport thrift.TTransport
+	replaceCounter metrics.Counter
 
-	// if expiration is zero, then the client will be kept open indefinetly.
-	expiration time.Time
+	// whole block guarded by lock
+	lock       sync.Mutex
+	client     thrift.TClient
+	transport  thrift.TTransport
+	expiration time.Time // if expiration is zero, then the client will be kept open indefinetly.
+	timer      *time.Timer
+	closed     bool
 }
 
 // Close implements Client interface.
 //
 // It calls underlying TTransport's Close function.
 func (c *ttlClient) Close() error {
+	c.lock.Lock()
+	defer c.lock.Unlock()
+	c.closed = true
+	if c.timer != nil {
+		c.timer.Stop()
+	}
 	return c.transport.Close()
 }
 
+func (c *ttlClient) Call(ctx context.Context, method string, args, result thrift.TStruct) (thrift.ResponseMeta, error) {
+	c.lock.Lock()
+	defer c.lock.Unlock()
+	return c.client.Call(ctx, method, args, result)
```

Comment:
it might be nice to check closed here too, to return a nice error message.

🔕 Also, it would be _great_ to annotate any error messages that get returned from this with the slug, since thrift doesn't do that it just gives bare IPs...